# Review of functions used in Disc1_intro

- `Function:` read.csv("NAMEOFFILE.csv")
    - Reads csv
- `Function:` ls()
    - Shows saved objects
- `Function:` head(DF_NAME, # of lines you want to display)
    - Show first lines of data frame
- `Function:` names(DF_NAME)
    - Shows column names
- `Function:` summary(DF_NAME)
    - Shows summary stats
- `Function:` mean(DF_NAME\$Col_Name)
    - Shows mean of certain column
- `Function:` sd(DF_NAME\$Col_name)
    - Shows standard deviation of certain column


# Introduction to DPLYR

## Table of contents

- [1.Basics](#1.-Basics)
- [2.DPLYR](#2.-DPLYR)
    - [2.1 Basics](#2.1-Basics)
    - [2.2 Select](#2.2-Select)
    - [2.3 Filter](#2.3-Filter)
    - [2.4 Arrange](#2.4-Arrange)
    - [2.5 Mutate](#2.5-Mutate)
    - [2.6 Summarise](#2.6-Summarise)
    - [2.7 Groupby](#2.7-Groupby)
    
- [3. Other DPLYR functions](#3.-Other-DPLYR-functions)
- [4. Review Questions](#4.-Review-Questions)

# 1. Basics

## 1.1 Find file [path](https://automatetheboringstuff.com/2e/chapter9/)

- This tells you what folder you are currently working in 


### This is an example of file paths on your computer
![title](Images/File_path.png)

### 1.1.1 Finding file paths on macs
- Copy the where section and paste should give you a file path
    - For example: Users/NAME/Desktop/R/filename.extension
![title](Images/File_path_mac.png)

### 1.1.2 Finding file path within R

`Function: read.csv("NAMEOFFILE.csv")`

In [1]:
path = getwd() #Sets current working directory
print(paste("The current working directory is: ", getwd(), sep=""))

[1] "The current working directory is: /Users/corybaird/Desktop/R/Disc2_DPLYR"


## 1.2 Import files
### 1.2.1 Import from: current working directory



#### 1.2.1.1 Lets import that file from an internal data set!


In [2]:
#Download internal data set: cars data
data("mtcars") 

#Show the first two lines of the data set
head(mtcars, 2)

,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Mazda RX4,21,6,160,110,3.9,2.620,16.46,0,1,4,4
Mazda RX4 Wag,21,6,160,110,3.9,2.875,17.02,0,1,4,4


##### 1.2.1.1.1 Write the internal data set to csv file

- `Function:` write.csv(DF_Name, "NAMEOFFILE.csv")
    - Writes data frame to csv file

In [3]:
write.csv(mtcars, "cars.csv") #Name the dataset "cars.csv" and save to current working directory

#### 1.2.1.2 Now import "cars.csv" which is now in our current directory

In [4]:
# Avoid the following mistake when importing data
# read.csv("cars.csv")

# Must add "YOURNAME =" to save the dataframe!
df = read.csv("cars.csv") #We often name our dataframes df_NAME

##### 1.2.1.2.1 Check the first two lines of the cars.csv file now imported as "df"

In [5]:
head(df,2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Mazda RX4,21,6,160,110,3.9,2.620,16.46,0,1,4,4
Mazda RX4 Wag,21,6,160,110,3.9,2.875,17.02,0,1,4,4


### 1.2.2  Import from: not in our current directory

We need to know the file path of the file!!

#### 1.2.2.1 Create new directory (folder) and place new data set

- You do not need to understand the following code
- It is a fancy way to create a new folder and saves a csv file inside that new folder
- We will import the csv file from that new folder



- `Function:` dir.create("Folder_name")
    - Creates new folder
- `Function:` setwd("FILEPATH")
    - Changes file path

In [6]:
dir.create('New_directory') #Creates new folder

setwd("New_directory") #Changes working directory

data("USArrests") #Imports internal data set

write.csv(USArrests, "arrests.csv") #Name the dataset "arrests.csv" and save to current working directory

setwd(path) #Returns to our original file path!!!

In [7]:
getwd() #Checks our path which is not in the folder we just created

[1] "/Users/corybaird/Desktop/R/Disc2_DPLYR"

#### 1.2.2.2 Import "arrests.csv" from the "New_directory" folder
- We can do this even from our original file path position because we added the name of the folder before the file
    - read.csv("FOLDER_NAME/FILE_NAME.csv")

In [8]:
df_arrests = read.csv("New_directory/arrests.csv")

In [9]:
head(df_arrests,2)

X,Murder,Assault,UrbanPop,Rape
Alabama,13.2,236,58,21.2
Alaska,10.0,263,48,44.5


# 2. DPLYR

We will use cars data set we named df

## 2.1 Basics

### 2.1.1 Download and import dplyr package

#### 2.1.1.1 Downloading packages

- There are two ways to download packages
    - install.packages('Package_name')
    - Click the. "packages" tab in the lower right hand and install via r-studio
        - See screenshot
    
#### Note: You only need to install the package one time!
- So do not run the install.packages('Package_name') command more than once


![title](Images/Install_package.png)

In [11]:
#install.packages("dplyr") #Already installed!

library(dplyr) # Do not worry if you see red code those are just warning messages

### 2.2 Select
- Select columns

In [14]:
#We can list all the column names from a Data Frame
colnames(df)

[1] "X"    "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"  
[11] "gear" "carb"

### 2.2.1 Select one column

In [15]:
df %>% #Dataframe
select(X)%>% #Select column named X
head(3) #Shows only first 3 obs

X
Mazda RX4
Mazda RX4 Wag
Datsun 710


### 2.2.2 Select multiple columns

In [17]:
df %>% #Dataframe
select(cyl, disp, hp, wt)%>% #Select column named X
head(2) #Shows only first 2 obs

cyl,disp,hp,wt
6,160,110,2.620
6,160,110,2.875


### 2.2.3 QUESTION: Select columns and create new Data Frame
Hint: We can do this by create a new object 
    - NEW_NAME = 

### 2.2.4 Select by column name

#### 2.2.4.1 Select by name that starts with h

In [18]:
df %>% #Dataframe
select(starts_with('h'))%>% #Select column named X
head(2) #Shows only first 2 obs

hp
110
110


#### 2.2.4.2 Select by name that starts with t

In [20]:
df %>% #Dataframe
select(ends_with('t'))%>% #Select column named X
head(2) #Shows only first 2 obs

drat,wt
3.9,2.620
3.9,2.875


#### 2.2.4.3 Select by name that contains c

In [21]:
df %>% #Dataframe
select(contains('c'))%>% #Select column named X
head(2) #Shows only first 2 obs

cyl,qsec,carb
6,16.46,4
6,17.02,4


### 2.2.4 Pull: similar to select expect it only extracts values

In [23]:
df %>% #Dataframe
pull(var = cyl)

[1] 6 6 4 6 8 6 8 4 4 6 6 8 8 8 8 8 8 4 4 4 4 8 8 8 8 4 4 4 8 6 8 4

## 2.3 Filter 

In [24]:
head(df,2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Mazda RX4,21,6,160,110,3.9,2.620,16.46,0,1,4,4
Mazda RX4 Wag,21,6,160,110,3.9,2.875,17.02,0,1,4,4


### 2.3.1 Filter by 1 condition

In [25]:
df %>% 
filter(mpg>22)%>% #filter all observations over 25 mpg
head(2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Datsun 710,22.8,4,108.0,93,3.85,2.32,18.61,1,1,4,1
Merc 240D,24.4,4,146.7,62,3.69,3.19,20.00,1,0,4,2


### 2.3.2 Filter by 2 conditions

In [26]:
df %>% 
filter(mpg>22 & mpg<35) %>% head(2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Datsun 710,22.8,4,108.0,93,3.85,2.32,18.61,1,1,4,1
Merc 240D,24.4,4,146.7,62,3.69,3.19,20.00,1,0,4,2


### 2.3.3 Filter by gear not equal to 4

In [27]:
df %>% 
filter(gear!=4)%>% head(2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Hornet 4 Drive,21.4,6,258,110,3.08,3.215,19.44,1,0,3,1
Hornet Sportabout,18.7,8,360,175,3.15,3.440,17.02,0,0,3,2


### 2.3.4 QUESTION: Filter by vs=1

### 2.3.5 Filter by Mazda 
-grepl function

In [28]:
df %>% filter(grepl("Maz",X))

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Mazda RX4,21,6,160,110,3.9,2.620,16.46,0,1,4,4
Mazda RX4 Wag,21,6,160,110,3.9,2.875,17.02,0,1,4,4


## 2.4 Arrange

### 2.4.1 Arrange by column (Ascending)

In [29]:
df %>% 
arrange(mpg)%>% 
head(3)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Cadillac Fleetwood,10.4,8,472,205,2.93,5.250,17.98,0,0,3,4
Lincoln Continental,10.4,8,460,215,3.00,5.424,17.82,0,0,3,4
Camaro Z28,13.3,8,350,245,3.73,3.840,15.41,0,0,3,4


### 2.4.2 Arrange  (Descending)

In [30]:
df %>% 
arrange(desc(mpg))%>% 
head(3)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Toyota Corolla,33.9,4,71.1,65,4.22,1.835,19.90,1,1,4,1
Fiat 128,32.4,4,78.7,66,4.08,2.200,19.47,1,1,4,1
Honda Civic,30.4,4,75.7,52,4.93,1.615,18.52,1,1,4,2


## 2.5 Mutate 
- Create new column

### 2.5.1 Mutate

In [31]:
df %>% 
mutate(mpg_10 = mpg/10) %>% head(2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb,mpg_10
Mazda RX4,21,6,160,110,3.9,2.620,16.46,0,1,4,4,2.1
Mazda RX4 Wag,21,6,160,110,3.9,2.875,17.02,0,1,4,4,2.1


#### 2.5.1.1 Mutate: Various functions
1. Cumulative sum: cumsum(X)
2. Lag: lag(X)
3. percent_rank(X)

In [32]:
df %>% 
mutate(cumulative_wt = cumsum(wt), #cumulative
       lag_wt = lag(wt), #Opposite function: lead()
       Percent_wt = percent_rank(wt)
      ) %>% head(2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb,cumulative_wt,lag_wt,Percent_wt
Mazda RX4,21,6,160,110,3.9,2.620,16.46,0,1,4,4,2.620,NA,0.2580645
Mazda RX4 Wag,21,6,160,110,3.9,2.875,17.02,0,1,4,4,5.495,2.62,0.3548387


### 2.5.2 Transmute
- Same as mutate expect it drops other columns

In [33]:
df %>% 
transmute(mpg_10 = mpg/10)%>% head(2)

mpg_10
2.1
2.1


### 2.5.3 Mutate_all

- Mutate all selected columns by certain function
- This example shows log
    - funs(log(.))

In [35]:
df[, 2:5]%>% 
mutate_all(funs(log(.)))%>%
head(3)

mpg,cyl,disp,hp
3.044522,1.791759,5.075174,4.700480
3.044522,1.791759,5.075174,4.700480
3.126761,1.386294,4.682131,4.532599


### 2.5.4 Mutate_at

Mutates only at specified columns

In [36]:
df%>% 
mutate_at(vars(mpg, disp, hp), funs(log(.)))%>% 
head(2)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Mazda RX4,3.044522,6,5.075174,4.70048,3.9,2.620,16.46,0,1,4,4
Mazda RX4 Wag,3.044522,6,5.075174,4.70048,3.9,2.875,17.02,0,1,4,4


## 2.6 Summarise

In [37]:
df %>% 
summarise(mean_mpg = mean(mpg),
        median_mpg = median(mpg),
         sum_mpg = sum(mpg),
         count_mpg = n(),
         first_obs = first(mpg),
         last_obs = last(mpg),
         variance = var(mpg),
         sd = sd(mpg))

mean_mpg,median_mpg,sum_mpg,count_mpg,first_obs,last_obs,variance,sd
20.09062,19.2,642.9,32,21,21.4,36.3241,6.026948


## 2.7 Groupby 

In [38]:
df %>% 
group_by(gear) %>% 
summarise(mpg_by_gear = mean(mpg))

gear,mpg_by_gear
3,16.10667
4,24.53333
5,21.38000


# 3. Other DPLYR functions

## 3.1 Count

### 3.1.1 Count number of observations by gear

In [39]:
df %>% 
count(gear)

gear,n
3,15
4,12
5,5


## 3.2 Distinct

In [40]:
df %>% 
distinct(X)%>% head(3)

X
Mazda RX4
Mazda RX4 Wag
Datsun 710


## 3.3 Sample data

Used for train and test split in ML and NN

### 3.3.1 Sample 50% of the data

In [41]:
df %>% 
sample_frac(.5)%>% head(3)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Toyota Corolla,33.9,4,71.1,65,4.22,1.835,19.90,1,1,4,1
Pontiac Firebird,19.2,8,400.0,175,3.08,3.845,17.05,0,0,3,2
Merc 230,22.8,4,140.8,95,3.92,3.150,22.90,1,0,4,2


### 3.3.2 Sample 10 observations

In [42]:
df %>% 
sample_n(10)%>% head(3)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Hornet Sportabout,18.7,8,360.0,175,3.15,3.440,17.02,0,0,3,2
Toyota Corona,21.5,4,120.1,97,3.70,2.465,20.01,1,0,3,1
Honda Civic,30.4,4,75.7,52,4.93,1.615,18.52,1,1,4,2


## 3.4 Slice
Pick out rows by observation number

In [43]:
df %>% 
slice(10:12)

X,mpg,cyl,disp,hp,drat,wt,qsec,vs,am,gear,carb
Merc 280,19.2,6,167.6,123,3.92,3.44,18.3,1,0,4,4
Merc 280C,17.8,6,167.6,123,3.92,3.44,18.9,1,0,4,4
Merc 450SE,16.4,8,275.8,180,3.07,4.07,17.4,0,0,3,3


# 4. Review Questions

## Q.1. Show mean mpg by gear

## Q.2. Show all observations with: am=1 & vs=0

## Q.3 Create new column for Kilometers per gallon (mulitply mpg*1.6)

## Q.4 Show all observations with a gear = 4 but instead of showing the whole frame only showing cyl and disp columns